### PR TITLE
add model

### PR DESCRIPTION
**Pull request by patrickvonplaten**

### CODE REVIEW

File: docs/source/en/api/pipelines/if.mdx [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+from diffusers import IFInpaintingPipeline, IFInpaintingSuperResolutionPipeline
+
+pipe_1 = IFInpaintingPipeline(**pipe_1.components)
+pipe_2 = IFInpaintingSuperResolutionPipeline(**pipe_2.components)
+```
+
+### Optimizing for speed
+
+TODO:
```
Comment (patrickvonplaten) [inline docs/source/en/api/pipelines/if.mdx:242]: @williamberman could you take care of this

File: docs/source/en/api/pipelines/if.mdx [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+- use custom timesteps
+- reduce timesteps
+- leave everything on GPU
+- for image-to-image use weak strength
+- For future, `torch.compile`
+
+### Optimizing for memory
+
+TODO:
```
Comment (patrickvonplaten) [inline docs/source/en/api/pipelines/if.mdx:251]: @williamberman could you take care of this

File: src/diffusers/pipelines/deepfloyd_if/watermark.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class IFWatermarker(ModelMixin, ConfigMixin):
+    def __init__(self):
+        super().__init__()
+
+        self.register_buffer("watermark_image", torch.zeros((62, 62, 4)))
+        self.watermark_image_as_pil = None
+
+    def apply_watermark(self, images: List[PIL.Image.Image], sample_size):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/deepfloyd_if/watermark.py:19]:
```suggestion
    def forward(self, images: List[PIL.Image.Image], sample_size):
```

Comment (patrickvonplaten) [reply src/diffusers/pipelines/deepfloyd_if/watermark.py:19]: And apply to all pipelines - @williamberman could you take care of this?

File: docs/source/en/api/pipelines/overview.mdx [hunk 4]
```diff
@@ -51,6 +51,9 @@ available a colab notebook to directly try them out.
...
 | [ddpm](./ddpm) | [**Denoising Diffusion Probabilistic Models**](https://arxiv.org/abs/2006.11239) | Unconditional Image Generation |
 | [ddim](./ddim) | [**Denoising Diffusion Implicit Models**](https://arxiv.org/abs/2010.02502) | Unconditional Image Generation |
+| [if](./if) | [**IF**](TODO link) | Image Generation | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/deepfloyd_if_free_tier_google_colab.ipynb)
+| [if_img2img](./if) | [**IF**](TODO link) | Image-to-Image Generation | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/deepfloyd_if_free_tier_google_colab.ipynb)
+| [if_inpainting](./if) | [**IF**](TODO link) | Image-to-Image Generation | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/deepfloyd_if_free_tier_google_colab.ipynb)
```
Comment (patrickvonplaten) [inline docs/source/en/api/pipelines/overview.mdx:56]: @williamberman TODO

File: docs/source/en/index.mdx [hunk 5]
```diff
@@ -58,6 +58,9 @@ The library has three main components:
 | [dance_diffusion](./api/pipelines/dance_diffusion) | [Dance Diffusion](https://github.com/williamberman/diffusers.git) | Unconditional Audio Generation |
 | [ddpm](./api/pipelines/ddpm) | [Denoising Diffusion Probabilistic Models](https://arxiv.org/abs/2006.11239) | Unconditional Image Generation |
 | [ddim](./api/pipelines/ddim) | [Denoising Diffusion Implicit Models](https://arxiv.org/abs/2010.02502) | Unconditional Image Generation |
+| [if](./if) | [**IF**](TODO link) | Image Generation |
```
Comment (patrickvonplaten) [inline docs/source/en/index.mdx:61]: @williamberman TODO

File: docs/source/en/index.mdx [hunk 6]
```diff
@@ -58,6 +58,9 @@ The library has three main components:
...
+| [if_img2img](./if) | [**IF**](TODO link) | Image-to-Image Generation |
```
Comment (patrickvonplaten) [inline docs/source/en/index.mdx:62]: @williamberman TODO

File: docs/source/en/index.mdx [hunk 7]
```diff
@@ -58,6 +58,9 @@ The library has three main components:
...
+| [if_inpainting](./if) | [**IF**](TODO link) | Image-to-Image Generation |
```
Comment (patrickvonplaten) [inline docs/source/en/index.mdx:63]: @williamberman TODO

File: docs/source/en/api/pipelines/if.mdx [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+    original_image=original_image,
+    mask_image=mask_image,
+    prompt_embeds=prompt_embeds,
+    negative_prompt_embeds=negative_embeds,
+    output_type="pt",
+).images
+pt_to_pil(image)[0].save("./if_stage_II.png")
+
+# stage 3
+image = stage_3(prompt=prompt, image=image, noise_level=100).images[0]
```
Comment (patrickvonplaten) [inline docs/source/en/api/pipelines/if.mdx:236]: @williamberman we need to change to add watermark here

File: docs/source/en/api/pipelines/if.mdx [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+    image=image,
+    original_image=original_image,
+    prompt_embeds=prompt_embeds,
+    negative_prompt_embeds=negative_embeds,
+    output_type="pt",
+).images
+pt_to_pil(image)[0].save("./if_stage_II.png")
+
+# stage 3
+image = stage_3(prompt=prompt, image=image, noise_level=100).images[0]
```
Comment (patrickvonplaten) [inline docs/source/en/api/pipelines/if.mdx:162]: @williamberman we need to change to add watermark here

File: src/diffusers/pipelines/deepfloyd_if/watermark.py (hunk 3 above)
Comment (williamberman) [reply src/diffusers/pipelines/deepfloyd_if/watermark.py:19]: discussed on slack

File: docs/source/en/api/pipelines/overview.mdx [hunk 10]
```diff
@@ -51,6 +51,9 @@ available a colab notebook to directly try them out.
 | [dance_diffusion](./dance_diffusion) | [**Dance Diffusion**](https://github.com/williamberman/diffusers.git) | Unconditional Audio Generation |
 | [ddpm](./ddpm) | [**Denoising Diffusion Probabilistic Models**](https://arxiv.org/abs/2006.11239) | Unconditional Image Generation |
 | [ddim](./ddim) | [**Denoising Diffusion Implicit Models**](https://arxiv.org/abs/2010.02502) | Unconditional Image Generation |
+| [if](./if) | [**IF**](TODO link) | Image Generation | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/deepfloyd_if_free_tier_google_colab.ipynb)
```
[the remark this answers is not in the document]
Comment (williamberman) [reply docs/source/en/api/pipelines/overview.mdx]: https://github.com/huggingface/diffusers/pull/3230/commits/14ba8bc44e01cbdd7181ab4d2792b0370f6bcdcc

File: docs/source/en/api/pipelines/overview.mdx [hunk 11]
```diff
@@ -51,6 +51,9 @@ available a colab notebook to directly try them out.
...
+| [if_img2img](./if) | [**IF**](TODO link) | Image-to-Image Generation | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/deepfloyd_if_free_tier_google_colab.ipynb)
```
[the remark this answers is not in the document]
Comment (williamberman) [reply docs/source/en/api/pipelines/overview.mdx]: https://github.com/huggingface/diffusers/pull/3230/commits/14ba8bc44e01cbdd7181ab4d2792b0370f6bcdcc

File: docs/source/en/api/pipelines/overview.mdx (hunk 4 above)
Comment (williamberman) [reply docs/source/en/api/pipelines/overview.mdx]: https://github.com/huggingface/diffusers/pull/3230/commits/14ba8bc44e01cbdd7181ab4d2792b0370f6bcdcc

File: docs/source/en/index.mdx (hunk 5 above)
Comment (williamberman) [reply docs/source/en/index.mdx]: https://github.com/huggingface/diffusers/pull/3230/commits/14ba8bc44e01cbdd7181ab4d2792b0370f6bcdcc

File: docs/source/en/index.mdx (hunk 7 above)
Comment (williamberman) [reply docs/source/en/index.mdx]: https://github.com/huggingface/diffusers/pull/3230/commits/14ba8bc44e01cbdd7181ab4d2792b0370f6bcdcc

File: docs/source/en/api/pipelines/if.mdx (hunk 8 above)
Comment (williamberman) [reply docs/source/en/api/pipelines/if.mdx]: https://github.com/huggingface/diffusers/pull/3230/commits/7dd44ddf61a3b295f97953af0b93057c722ba04e

Comment (williamberman) [reply docs/source/en/api/pipelines/if.mdx]: added watermark and safetychecker

File: docs/source/en/api/pipelines/if.mdx (hunk 9 above)
Comment (williamberman) [reply docs/source/en/api/pipelines/if.mdx]: added watermark and safetychecker https://github.com/huggingface/diffusers/pull/3230/commits/7dd44ddf61a3b295f97953af0b93057c722ba04e

File: docs/source/en/api/pipelines/if.mdx [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+pt_to_pil(image)[0].save("./if_stage_I.png")
+
+# stage 2
+image = stage_2(
+    image=image, prompt_embeds=prompt_embeds, negative_prompt_embeds=negative_embeds, output_type="pt"
+).images
+pt_to_pil(image)[0].save("./if_stage_II.png")
+
+# stage 3
+image = stage_3(prompt=prompt, image=image, noise_level=100).images[0]
```
[the remark this answers is not in the document]
Comment (williamberman) [reply docs/source/en/api/pipelines/if.mdx]: added watermark and safetychecker https://github.com/huggingface/diffusers/pull/3230/commits/7dd44ddf61a3b295f97953af0b93057c722ba04e

File: docs/source/en/api/pipelines/if.mdx (hunk 1 above)
Comment (williamberman) [reply docs/source/en/api/pipelines/if.mdx]: https://github.com/huggingface/diffusers/pull/3230/commits/e02343f68fff2ad98d402ff2425901fe36d194dd

File: docs/source/en/api/pipelines/if.mdx (hunk 2 above)
Comment (williamberman) [reply docs/source/en/api/pipelines/if.mdx]: https://github.com/huggingface/diffusers/pull/3230/commits/f78c19ef6cf9015c13e1c65a01fbade02927626e

**Pull request merged**